# panicker3000/grpc/grpc_client.py
"""
gRPC –∫–ª–∏–µ–Ω—Ç –¥–ª—è –≤–∑–∞–∏–º–æ–¥–µ–π—Å—Ç–≤–∏—è —Å —Å–µ—Ä–≤–µ—Ä–æ–º –ü–∞–Ω–∏–∫—ë—Ä–∞.
–ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è Telegram-–±–æ—Ç–æ–º –≤–º–µ—Å—Ç–æ –ø—Ä—è–º—ã—Ö –≤—ã–∑–æ–≤–æ–≤ PanicDetector.
"""

# ============================================================================
# –ò–ú–ü–û–†–¢–´
# ============================================================================
import grpc
import logging
from typing import List, Dict, Any, Optional, Union
from datetime import datetime, timedelta
import time

# ============================================================================
# PYDANTIC SCHEMAS IMPORT
# ============================================================================
try:
    # –ü—Ä–æ–±—É–µ–º –∏–º–ø–æ—Ä—Ç —Å –ø—Ä–µ—Ñ–∏–∫—Å–æ–º panicker3000 (—Å—Ç—Ä—É–∫—Ç—É—Ä–∞ –ø–∞–∫–µ—Ç–∞)
    from panicker3000.utils.schemas import PanicSignal, validate_panic_signal

    PYDANTIC_AVAILABLE = True
except ImportError as e:
    # –ï—Å–ª–∏ –∏–º–ø–æ—Ä—Ç —Å –ø—Ä–µ—Ñ–∏–∫—Å–æ–º –Ω–µ —Å—Ä–∞–±–æ—Ç–∞–ª, –ø—Ä–æ–±—É–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π –ø—É—Ç—å
    try:
        from utils.schemas import PanicSignal, validate_panic_signal

        PYDANTIC_AVAILABLE = True
    except ImportError as e2:
        logging.warning(f"Pydantic schemas –Ω–µ–¥–æ—Å—Ç—É–ø–Ω—ã: {e2}")
        PanicSignal = None
        validate_panic_signal = None
        PYDANTIC_AVAILABLE = False

try:
    # gRPC –º–æ–¥—É–ª–∏ –Ω–∞—Ö–æ–¥—è—Ç—Å—è –≤ –ø–æ–¥–¥–∏—Ä–µ–∫—Ç–æ—Ä–∏–∏ grpc/proto/generated/
    import sys
    import os

    # 1. –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã–º proto —Ñ–∞–π–ª–∞–º
    current_dir = os.path.dirname(os.path.abspath(__file__))  # –ø–∞–ø–∫–∞ grpc
    proto_generated_path = os.path.join(current_dir, 'proto', 'generated')

    if proto_generated_path not in sys.path:
        sys.path.insert(0, proto_generated_path)

    # 2. –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º –º–æ–¥—É–ª–∏
    import panicker_pb2
    import panicker_pb2_grpc

except ImportError as e:
    print(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: –Ω–µ —É–¥–∞–ª–æ—Å—å –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞—Ç—å gRPC –º–æ–¥—É–ª–∏: {e}")
    print(f"–ü—Ä–æ–≤–µ—Ä—å—Ç–µ –Ω–∞–ª–∏—á–∏–µ —Ñ–∞–π–ª–æ–≤ –≤ {proto_generated_path}:")
    print(f"  - panicker_pb2.py")
    print(f"  - panicker_pb2_grpc.py")
    # –°–æ–∑–¥–∞—ë–º –∑–∞–≥–ª—É—à–∫–∏ —á—Ç–æ–±—ã –∫–æ–¥ –º–æ–≥ –∑–∞–ø—É—Å—Ç–∏—Ç—å—Å—è
    panicker_pb2 = None
    panicker_pb2_grpc = None
    print("‚ö†Ô∏è –°–æ–∑–¥–∞–Ω—ã –∑–∞–≥–ª—É—à–∫–∏ –¥–ª—è gRPC –º–æ–¥—É–ª–µ–π")

logger = logging.getLogger(__name__)


# ============================================================================
# –ö–õ–ê–°–° GrpcClient
# ============================================================================
class GrpcClient:
    """–ö–ª–∏–µ–Ω—Ç –¥–ª—è –≤–∑–∞–∏–º–æ–¥–µ–π—Å—Ç–≤–∏—è —Å gRPC —Å–µ—Ä–≤–µ—Ä–æ–º"""

    # ------------------------------------------------------------------------
    # –ò–ù–ò–¶–ò–ê–õ–ò–ó–ê–¶–ò–Ø
    # ------------------------------------------------------------------------
    def __init__(self, host: str = 'localhost', port: int = 50051):
        """
        –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è gRPC –∫–ª–∏–µ–Ω—Ç–∞

        Args:
            host: –•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞
            port: –ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞
        """
        self.host = host
        self.port = port
        self.channel = grpc.insecure_channel(f'{host}:{port}')

        # –°–æ–∑–¥–∞—ë–º –∑–∞–≥–ª—É—à–∫–∏ –¥–ª—è –≤—Å–µ—Ö —Å–µ—Ä–≤–∏—Å–æ–≤
        self.panicker_stub = panicker_pb2_grpc.PanickerServiceStub(self.channel)
        self.market_stub = panicker_pb2_grpc.MarketDataServiceStub(self.channel)
        self.signals_stub = panicker_pb2_grpc.SignalsServiceStub(self.channel)

        logger.info(f"gRPC –∫–ª–∏–µ–Ω—Ç –ø–æ–¥–∫–ª—é—á—ë–Ω –∫ {host}:{port}")

    # ------------------------------------------------------------------------
    # –ú–ï–¢–û–î–´ PanickerService
    # ------------------------------------------------------------------------
    def get_overheat_index(self, ticker: str) -> Dict[str, Any]:
        """
        –ü–æ–ª—É—á–∏—Ç—å –∏–Ω–¥–µ–∫—Å –ø–µ—Ä–µ–≥—Ä–µ–≤–∞ –¥–ª—è –∞–∫—Ü–∏–∏

        Args:
            ticker: –¢–∏–∫–µ—Ä –∞–∫—Ü–∏–∏

        Returns:
            –°–ª–æ–≤–∞—Ä—å —Å –¥–∞–Ω–Ω—ã–º–∏ –∏–Ω–¥–µ–∫—Å–∞ –ø–µ—Ä–µ–≥—Ä–µ–≤–∞
        """
        logger.info(f"–ó–∞–ø—Ä–æ—Å –∏–Ω–¥–µ–∫—Å–∞ –ø–µ—Ä–µ–≥—Ä–µ–≤–∞ –¥–ª—è {ticker}")

        try:
            response = self.panicker_stub.GetOverheatIndex(
                panicker_pb2.Ticker(symbol=ticker)
            )

            return {
                'ticker': response.ticker,
                'overheat_percentage': response.overheat_percentage,
                'current_rsi': response.current_rsi,
                'volume_ratio': response.volume_ratio,
                'last_signal_time': response.last_signal_time,
                'last_signal_level': self._convert_level_from_proto(response.last_signal_level)
            }

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ –∏–Ω–¥–µ–∫—Å–∞ –ø–µ—Ä–µ–≥—Ä–µ–≤–∞ {ticker}: {e}")
            return self._get_default_overheat_response(ticker)
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ –∏–Ω–¥–µ–∫—Å–∞ –ø–µ—Ä–µ–≥—Ä–µ–≤–∞ {ticker}: {e}")
            return self._get_default_overheat_response(ticker)

    def scan_tickers(self, tickers: List[str], real_time: bool = True) -> List[Union[Dict[str, Any], PanicSignal]]:
        """
        –°–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ —Ç–∏–∫–µ—Ä–æ–≤ –Ω–∞ –Ω–∞–ª–∏—á–∏–µ –ø–∞–Ω–∏–∫–∏/–∂–∞–¥–Ω–æ—Å—Ç–∏

        Returns:
            –°–ø–∏—Å–æ–∫ PanicSignal –º–æ–¥–µ–ª–µ–π –∏–ª–∏ —Å–ª–æ–≤–∞—Ä–µ–π (–µ—Å–ª–∏ Pydantic –Ω–µ–¥–æ—Å—Ç—É–ø–µ–Ω)
        """
        logger.info(f"–°–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ {len(tickers)} —Ç–∏–∫–µ—Ä–æ–≤ (—Ä–µ–∂–∏–º: {'real-time' if real_time else 'historical'})")

        try:
            ticker_objs = [panicker_pb2.Ticker(symbol=t) for t in tickers]
            request = panicker_pb2.ScanRequest(tickers=ticker_objs, real_time=real_time)

            response = self.panicker_stub.ScanTickers(request)

            signals = []
            for signal in response.signals:
                converted = self._convert_signal_from_proto(signal)
                signals.append(converted)

            # –õ–æ–≥–∏—Ä—É–µ–º —Ç–∏–ø –≤–æ–∑–≤—Ä–∞—â–∞–µ–º—ã—Ö –¥–∞–Ω–Ω—ã—Ö
            if signals and PYDANTIC_AVAILABLE and isinstance(signals[0], PanicSignal):
                logger.info(f"‚úÖ –ù–∞–π–¥–µ–Ω–æ {len(signals)} Pydantic —Å–∏–≥–Ω–∞–ª–æ–≤ –∏–∑ {response.total_scanned} —Ç–∏–∫–µ—Ä–æ–≤")
            else:
                logger.info(f"‚ö†Ô∏è –ù–∞–π–¥–µ–Ω–æ {len(signals)} dict —Å–∏–≥–Ω–∞–ª–æ–≤ –∏–∑ {response.total_scanned} —Ç–∏–∫–µ—Ä–æ–≤")

            return signals

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–∏: {e}")
            return []
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–∏: {e}")
            return []

    def get_signal_history(self, ticker: str, days_back: int = 7) -> List[Dict[str, Any]]:
        """
        –ü–æ–ª—É—á–∏—Ç—å –∏—Å—Ç–æ—Ä–∏—é —Å–∏–≥–Ω–∞–ª–æ–≤ –¥–ª—è —Ç–∏–∫–µ—Ä–∞

        Args:
            ticker: –¢–∏–∫–µ—Ä –∞–∫—Ü–∏–∏
            days_back: –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –¥–Ω–µ–π –∏—Å—Ç–æ—Ä–∏–∏

        Returns:
            –°–ø–∏—Å–æ–∫ –∏—Å—Ç–æ—Ä–∏—á–µ—Å–∫–∏—Ö —Å–∏–≥–Ω–∞–ª–æ–≤
        """
        logger.info(f"–ó–∞–ø—Ä–æ—Å –∏—Å—Ç–æ—Ä–∏–∏ —Å–∏–≥–Ω–∞–ª–æ–≤ –¥–ª—è {ticker} –∑–∞ {days_back} –¥–Ω–µ–π")

        try:
            end_date = datetime.now().isoformat()
            start_date = (datetime.now() - timedelta(days=days_back)).isoformat()

            request = panicker_pb2.HistoryRequest(
                ticker=ticker,
                days_back=days_back,
                limit=100
            )

            response = self.panicker_stub.GetSignalHistory(request)

            signals = []
            for signal in response.signals:
                signals.append(self._convert_signal_from_proto(signal))

            logger.info(f"–ü–æ–ª—É—á–µ–Ω–æ {len(signals)} –∏—Å—Ç–æ—Ä–∏—á–µ—Å–∫–∏—Ö —Å–∏–≥–Ω–∞–ª–æ–≤")
            return signals

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ –∏—Å—Ç–æ—Ä–∏–∏: {e}")
            return []
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ –∏—Å—Ç–æ—Ä–∏–∏: {e}")
            return []

    # ------------------------------------------------------------------------
    # –ú–ï–¢–û–î–´ MarketDataService
    # ------------------------------------------------------------------------
    def get_current_price(self, ticker: str) -> Optional[float]:
        """
        –ü–æ–ª—É—á–∏—Ç—å —Ç–µ–∫—É—â—É—é —Ü–µ–Ω—É —Ç–∏–∫–µ—Ä–∞

        Args:
            ticker: –¢–∏–∫–µ—Ä –∞–∫—Ü–∏–∏

        Returns:
            –¢–µ–∫—É—â–∞—è —Ü–µ–Ω–∞ –∏–ª–∏ None –ø—Ä–∏ –æ—à–∏–±–∫–µ
        """
        logger.info(f"–ó–∞–ø—Ä–æ—Å —Ç–µ–∫—É—â–µ–π —Ü–µ–Ω—ã –¥–ª—è {ticker}")

        try:
            request = panicker_pb2.PriceRequest(tickers=[ticker])
            response = self.market_stub.GetCurrentPrices(request)

            price = response.prices.get(ticker)
            if price:
                return price
            else:
                logger.warning(f"–¶–µ–Ω–∞ –¥–ª—è {ticker} –Ω–µ –Ω–∞–π–¥–µ–Ω–∞ –≤ –æ—Ç–≤–µ—Ç–µ")
                return None

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Ü–µ–Ω—ã {ticker}: {e}")
            return None
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Ü–µ–Ω—ã {ticker}: {e}")
            return None

    def get_candles(self, ticker: str, interval: str = 'min5', count: int = 100) -> List[Dict[str, Any]]:
        """
        –ü–æ–ª—É—á–∏—Ç—å —Å–≤–µ—á–∏ –¥–ª—è —Ç–∏–∫–µ—Ä–∞

        Args:
            ticker: –¢–∏–∫–µ—Ä –∞–∫—Ü–∏–∏
            interval: –ò–Ω—Ç–µ—Ä–≤–∞–ª —Å–≤–µ—á–µ–π
            count: –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ —Å–≤–µ—á–µ–π

        Returns:
            –°–ø–∏—Å–æ–∫ —Å–≤–µ—á–µ–π
        """
        logger.info(f"–ó–∞–ø—Ä–æ—Å —Å–≤–µ—á–µ–π –¥–ª—è {ticker}, –∏–Ω—Ç–µ—Ä–≤–∞–ª {interval}, –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ {count}")

        try:
            request = panicker_pb2.CandleRequest(
                ticker=ticker,
                interval=interval,
                count=count
            )

            response = self.market_stub.GetCandles(request)

            candles = []
            for candle in response.candles:
                candles.append({
                    'ticker': candle.ticker,
                    'open': candle.open,
                    'high': candle.high,
                    'low': candle.low,
                    'close': candle.close,
                    'volume': candle.volume,
                    'timestamp': candle.timestamp,
                    'interval': candle.interval
                })

            logger.info(f"–ü–æ–ª—É—á–µ–Ω–æ {len(candles)} —Å–≤–µ—á–µ–π")
            return candles

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Å–≤–µ—á–µ–π {ticker}: {e}")
            return []
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Å–≤–µ—á–µ–π {ticker}: {e}")
            return []

    # ------------------------------------------------------------------------
    # –ú–ï–¢–û–î–´ SignalsService
    # ------------------------------------------------------------------------
    def get_top_signals(self, period: str = 'today', limit: int = 5) -> List[Dict[str, Any]]:
        """
        –ü–æ–ª—É—á–∏—Ç—å —Ç–æ–ø —Å–∏–≥–Ω–∞–ª–æ–≤ –∑–∞ –ø–µ—Ä–∏–æ–¥

        Args:
            period: –ü–µ—Ä–∏–æ–¥ ('today', 'week', 'month')
            limit: –ú–∞–∫—Å–∏–º–∞–ª—å–Ω–æ–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ —Å–∏–≥–Ω–∞–ª–æ–≤

        Returns:
            –°–ø–∏—Å–æ–∫ —Ç–æ–ø —Å–∏–≥–Ω–∞–ª–æ–≤
        """
        logger.info(f"–ó–∞–ø—Ä–æ—Å —Ç–æ–ø-{limit} —Å–∏–≥–Ω–∞–ª–æ–≤ –∑–∞ –ø–µ—Ä–∏–æ–¥ {period}")

        try:
            request = panicker_pb2.TopRequest(period=period, limit=limit)
            response = self.signals_stub.GetTopSignals(request)

            signals = []
            for signal in response.top_signals:
                signals.append(self._convert_signal_from_proto(signal))

            logger.info(f"–ü–æ–ª—É—á–µ–Ω–æ {len(signals)} —Ç–æ–ø —Å–∏–≥–Ω–∞–ª–æ–≤")
            return signals

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Ç–æ–ø —Å–∏–≥–Ω–∞–ª–æ–≤: {e}")
            return []
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Ç–æ–ø —Å–∏–≥–Ω–∞–ª–æ–≤: {e}")
            return []

    def ignore_ticker(self, ticker: str, duration_hours: int = 2) -> bool:
        """
        –ò–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞—Ç—å —Ç–∏–∫–µ—Ä –Ω–∞ —É–∫–∞–∑–∞–Ω–Ω–æ–µ –≤—Ä–µ–º—è

        Args:
            ticker: –¢–∏–∫–µ—Ä –∞–∫—Ü–∏–∏
            duration_hours: –î–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –∏–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞–Ω–∏—è –≤ —á–∞—Å–∞—Ö

        Returns:
            True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ, False –µ—Å–ª–∏ –æ—à–∏–±–∫–∞
        """
        logger.info(f"–ò–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞–Ω–∏–µ {ticker} –Ω–∞ {duration_hours} —á–∞—Å–æ–≤")

        try:
            request = panicker_pb2.IgnoreRequest(
                ticker=ticker,
                duration_hours=duration_hours
            )

            response = self.signals_stub.IgnoreTicker(request)

            if response.success:
                logger.info(f"–¢–∏–∫–µ—Ä {ticker} –∏–≥–Ω–æ—Ä–∏—Ä—É–µ—Ç—Å—è –¥–æ {response.ignored_until}")
                return True
            else:
                logger.warning(f"–ù–µ —É–¥–∞–ª–æ—Å—å –∏–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞—Ç—å {ticker}")
                return False

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∏–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞–Ω–∏–∏ {ticker}: {e}")
            return False
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∏–≥–Ω–æ—Ä–∏—Ä–æ–≤–∞–Ω–∏–∏ {ticker}: {e}")
            return False

    # ------------------------------------------------------------------------
    # –í–°–ü–û–ú–û–ì–ê–¢–ï–õ–¨–ù–´–ï –ú–ï–¢–û–î–´
    # ------------------------------------------------------------------------
    def _convert_signal_from_proto(self, signal) -> Union[Dict[str, Any], PanicSignal]:
        """–ö–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏—è —Å–∏–≥–Ω–∞–ª–∞ –∏–∑ proto –≤ Pydantic –º–æ–¥–µ–ª—å (–∏–ª–∏ —Å–ª–æ–≤–∞—Ä—å –ø—Ä–∏ –æ—à–∏–±–∫–µ)"""
        level_map = {
            panicker_pb2.PanicSignal.STRONG: 'üî¥ –°–ò–õ–¨–ù–´–ô',
            panicker_pb2.PanicSignal.MODERATE: 'üü° –•–û–†–û–®–ò–ô',
            panicker_pb2.PanicSignal.URGENT: '‚ö™ –°–†–û–ß–ù–´–ô',
            panicker_pb2.PanicSignal.IGNORE: '‚ùå –ò–ì–ù–û–†–ò–†–û–í–ê–¢–¨'
        }

        signal_type_map = {
            panicker_pb2.PanicSignal.PANIC: '–ü–ê–ù–ò–ö–ê',
            panicker_pb2.PanicSignal.GREED: '–ñ–ê–î–ù–û–°–¢–¨',
            panicker_pb2.PanicSignal.NEUTRAL: '–ù–ï–ô–¢–†–ê–õ–¨–ù–û'
        }

        # –ë–∞–∑–æ–≤—ã–µ –ø–æ–ª—è
        result = {
            'ticker': signal.ticker,
            'signal_type': signal_type_map.get(signal.signal_type, '–ù–ï–ò–ó–í–ï–°–¢–ù–û'),
            'level': level_map.get(signal.level, '–ù–ï–ò–ó–í–ï–°–¢–ù–û'),
            'rsi_14': signal.rsi_14,
            'rsi_7': signal.rsi_7,
            'rsi_21': signal.rsi_21,
            'volume_ratio': signal.volume_ratio,
            'current_price': signal.current_price,
            'price': signal.current_price,  # –î–ª—è —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç–∏
            'detected_at': signal.detected_at,
            'timestamp': datetime.now().isoformat(),
        }

        # –î–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω—ã–µ –ø–æ–ª—è (–µ—Å–ª–∏ –µ—Å—Ç—å)
        if hasattr(signal, 'risk_metric'):
            result['risk_metric'] = signal.risk_metric
        else:
            result['risk_metric'] = 0.0

        if hasattr(signal, 'interpretation'):
            result['interpretation'] = signal.interpretation
        else:
            result['interpretation'] = ''

        # –ö–ª–∞—Å—Ç–µ—Ä—ã –æ–±—ä—ë–º–∞ (–µ—Å–ª–∏ –µ—Å—Ç—å)
        result['volume_clusters'] = []
        if hasattr(signal, 'volume_clusters'):
            for cluster in signal.volume_clusters:
                result['volume_clusters'].append({
                    'price_level': cluster.price_level,
                    'volume_percentage': cluster.volume_percentage,
                    'role': cluster.role
                })

        # ========================================================================
        # –ü–´–¢–ê–ï–ú–°–Ø –°–û–ó–î–ê–¢–¨ PYDANTIC –ú–û–î–ï–õ–¨
        # ========================================================================
        if PYDANTIC_AVAILABLE and validate_panic_signal:
            try:
                is_valid, pydantic_signal, error = validate_panic_signal(result)
                if is_valid and pydantic_signal:
                    logger.info(f"‚úÖ –°–æ–∑–¥–∞–Ω–∞ Pydantic –º–æ–¥–µ–ª—å –¥–ª—è {signal.ticker}")
                    return pydantic_signal  # –í–æ–∑–≤—Ä–∞—â–∞–µ–º PanicSignal
                else:
                    logger.warning(f"‚ö†Ô∏è –í–∞–ª–∏–¥–∞—Ü–∏—è –Ω–µ –ø—Ä–æ—à–ª–∞ –¥–ª—è {signal.ticker}: {error}")
                    return result  # –í–æ–∑–≤—Ä–∞—â–∞–µ–º —Å–ª–æ–≤–∞—Ä—å –∫–∞–∫ –∑–∞–ø–∞—Å–Ω–æ–π –≤–∞—Ä–∏–∞–Ω—Ç
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ —Å–æ–∑–¥–∞–Ω–∏—è Pydantic –º–æ–¥–µ–ª–∏: {e}")
                return result  # –í–æ–∑–≤—Ä–∞—â–∞–µ–º —Å–ª–æ–≤–∞—Ä—å

        # –ï—Å–ª–∏ Pydantic –Ω–µ–¥–æ—Å—Ç—É–ø–µ–Ω, –≤–æ–∑–≤—Ä–∞—â–∞–µ–º —Å–ª–æ–≤–∞—Ä—å
        logger.info(f"‚ö†Ô∏è Pydantic –Ω–µ–¥–æ—Å—Ç—É–ø–µ–Ω, –≤–æ–∑–≤—Ä–∞—â–∞–µ–º dict –¥–ª—è {signal.ticker}")
        return result

    def get_stats(self, days: int = 7) -> Dict[str, Any]:
        logger.info(f"–ó–∞–ø—Ä–æ—Å —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏ –∑–∞ {days} –¥–Ω–µ–π")

        try:
            request = panicker_pb2.StatsRequest(days=days)
            response = self.signals_stub.GetStats(request)

            return {
                'total_signals': response.total_signals,
                'strong_signals': response.strong_signals,
                'moderate_signals': response.moderate_signals,
                'urgent_signals': response.urgent_signals,
                'most_active_ticker': response.most_active_ticker,
                'most_active_count': response.most_active_count,
                'most_calm_ticker': response.most_calm_ticker,
                'most_calm_count': response.most_calm_count,
                'market_tension': response.market_tension
            }

        except grpc.RpcError as e:
            logger.error(f"gRPC –æ—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏: {e}")
            raise
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–ø—Ä–æ—Å–µ —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏: {e}")
            raise

    def _convert_level_from_proto(self, level_proto: int) -> str:
        """–ö–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏—è —É—Ä–æ–≤–Ω—è –∏–∑ proto –≤ —Å—Ç—Ä–æ–∫—É"""
        level_map = {
            panicker_pb2.PanicSignal.STRONG: 'üî¥ –°–ò–õ–¨–ù–´–ô',
            panicker_pb2.PanicSignal.MODERATE: 'üü° –•–û–†–û–®–ò–ô',
            panicker_pb2.PanicSignal.URGENT: '‚ö™ –°–†–û–ß–ù–´–ô',
            panicker_pb2.PanicSignal.IGNORE: '‚ùå –ò–ì–ù–û–†–ò–†–û–í–ê–¢–¨'
        }
        return level_map.get(level_proto, '–ù–ï–ò–ó–í–ï–°–¢–ù–û')

    def _get_default_overheat_response(self, ticker: str) -> Dict[str, Any]:
        """–û—Ç–≤–µ—Ç –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é –ø—Ä–∏ –æ—à–∏–±–∫–µ"""
        return {
            'ticker': ticker,
            'overheat_percentage': None,
            'current_rsi': None,
            'volume_ratio': None,
            'last_signal_time': '',
            'last_signal_level': '–û–®–ò–ë–ö–ê –ü–û–î–ö–õ–Æ–ß–ï–ù–ò–Ø'
        }

    def _get_default_stats_response(self) -> Dict[str, Any]:
        """–û—Ç–≤–µ—Ç –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é –ø—Ä–∏ –æ—à–∏–±–∫–µ –ø–æ–ª—É—á–µ–Ω–∏—è —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏"""
        return {
            'total_signals': None,
            'strong_signals': None,
            'moderate_signals': None,
            'urgent_signals': None,
            'most_active_ticker': None,
            'most_active_count': None,
            'most_calm_ticker': None,
            'most_calm_count': None,
            'market_tension': None
        }

    # ------------------------------------------------------------------------
    # –ó–ê–ö–†–´–¢–ò–ï –°–û–ï–î–ò–ù–ï–ù–ò–Ø
    # ------------------------------------------------------------------------
    def close(self):
        """–ó–∞–∫—Ä—ã—Ç—å —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–µ —Å —Å–µ—Ä–≤–µ—Ä–æ–º"""
        if self.channel:
            self.channel.close()
            logger.info("–°–æ–µ–¥–∏–Ω–µ–Ω–∏–µ —Å gRPC —Å–µ—Ä–≤–µ—Ä–æ–º –∑–∞–∫—Ä—ã—Ç–æ")


# ============================================================================
# –ì–õ–û–ë–ê–õ–¨–ù–´–ô –≠–ö–ó–ï–ú–ü–õ–Ø–† –ö–õ–ò–ï–ù–¢–ê
# ============================================================================
_client_instance: Optional[GrpcClient] = None


def get_grpc_client() -> GrpcClient:
    """–ü–æ–ª—É—á–∏—Ç—å –≥–ª–æ–±–∞–ª—å–Ω—ã–π —ç–∫–∑–µ–º–ø–ª—è—Ä gRPC –∫–ª–∏–µ–Ω—Ç–∞ (—Å–∏–Ω–≥–ª—Ç–æ–Ω)"""
    global _client_instance
    if _client_instance is None:
        _client_instance = GrpcClient()
    return _client_instance

